fix incorrect operator lookup for postincrement

maxbachmann opened the pull request:
Fixes #4535

da-woods commented:
In principle this might break some existing code (if they've decided to only wrap one of the operators). I recommend:

1. Add a sentence or two to https://github.com/cython/cython/blob/master/docs/src/userguide/migrating_to_cy30.rst
2. Would it be possible to customize the error message slightly to say something like "wrap c++ operator++(int)" just to help people identify what they need to change?

Point 2 is only if it's an easy change!

maxbachmann commented:
> In principle this might break some existing code (if they've decided to only wrap one of the operators)

This breaks the current iterator implementation, which only specifies the preincrement operator. I added the postincrement operators in https://github.com/cython/cython/pull/4528.

> Add a sentence or two to https://github.com/cython/cython/blob/master/docs/src/userguide/migrating_to_cy30.rst

I will update the migration guide

> Would it be possible to customize the error message slightly to say something like "wrap c++ operator++(int)" just to help people identify what they need to change?

I updated the error message to:
```
Invalid operand type for '++'. Wrap Foo::operator++(int)
Invalid operand type for '++'. Wrap Foo::operator++()
```

da-woods commented:
Thanks - it's fairly minor breakage and should be easy for users to fix, but hopefully giving them a little hint will make it easier to work out what to do

maxbachmann commented:
I just noted that this does not work for nonmember operators yet.

maxbachmann commented:
It would be good to fix https://github.com/cython/cython/issues/4539 first.

scoder commented:
>> Would it be possible to customize the error message slightly to say something like "wrap c++ operator++(int)" just to help people identify what they need to change?
> 
> I updated the error message to:
> 
> ```
> Invalid operand type for '++'. Wrap Foo::operator++(int)
> Invalid operand type for '++'. Wrap Foo::operator++()
> ```

It's not clear even to me what "wrap" means here without looking through the code. Did you mean to ask users to add these operators to their external C++ class declarations? Or to correct the existing declarations?

maxbachmann commented:
> It's not clear even to me what "wrap" means here without looking through the code. Did you mean to ask users to add these operators to their external C++ class declarations? Or to correct the existing declarations?

I think I will fix this to be closer to the error message of gcc, which gives the following errors in this case:
```
error: no 'operator++(int)' declared for postfix '++'
error: no match for 'operator++' (operand type is 'example')
```

maxbachmann commented:
@scoder @da-woods I completely forgot about this PR of mine. I updated the error message and added a paragraph to the Cython 3 migration guide. It would be great to get this in before Cython 3, since this is a least a slightly breaking change.

da-woods commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -10534,7 +10534,10 @@ def type_error(self):
     def analyse_cpp_operation(self, env, overload_check=True):
-        entry = env.lookup_operator(self.operator, [self.operand])
+        operand_types = [self.operand.type]
+        if not getattr(self, 'is_prefix', True):
```
Normally when I try to sneak in `getattr` I don't succeed. Therefore I'm going to propose a slightly different approach

```suggestion
        if self.is_inc_dec_op and not self.is_prefix:
```

maxbachmann replied:
just out of interest: what is the problem with getattr? Is this just a stylistic preference, or are the any problems with it?

da-woods replied:
Mostly stylistic I think. In fairness, this version (or a similar one with is instance) you immediately know it refers to a postfix increment or decrement operator because it's all there in the if statement. While with `getattr` you have to reason about why it might not be set, and why `True` is a suitable default value. So I think it is clearer

da-woods commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -10548,7 +10551,12 @@ def analyse_cpp_operation(self, env, overload_check=True):
             self.exception_check = ''
             self.exception_value = ''
-        cpp_type = self.operand.type.find_cpp_operation_type(self.operator)
+        if not getattr(self, 'is_prefix', True):
```
```suggestion
        if self.is_inc_dec_op and not self.is_prefix:
```

da-woods commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -10691,6 +10699,16 @@ def calculate_result_code(self):
 class DecrementIncrementNode(CUnopNode):
     #  unary ++/-- operator
 
+    def type_error(self):
```
```suggestion
    is_inc_dec_op = True

    def type_error(self):
```

The final part of this change is that `UnopNode` needs to have `is_inc_dec_op = False` but I can't suggest that through the github interface

da-woods commented on docs/src/userguide/migrating_to_cy30.rst:
```diff
@@ -226,3 +226,16 @@ To make it easier to handle cases where your interpretation of type
+
+Cython 3 differentiates between pre/post-increment and pre/pos-decrement
+operators (Cython 0.29 implemented both as predecrement operator).
+This only has an effect when using ``operator.postdecrement`` / ``operator.postincrement``.
```
```suggestion
This only has an effect when using ``cython.operator.postdecrement`` / ``cython.operator.postincrement``.
```

?

da-woods commented on docs/src/userguide/migrating_to_cy30.rst:
```diff
@@ -226,3 +226,16 @@ To make it easier to handle cases where your interpretation of type
+C++ postincrement/postdecrement operator
+========================================
+
+Cython 3 differentiates between pre/post-increment and pre/pos-decrement
```
```suggestion
Cython 3 differentiates between pre/post-increment and pre/post-decrement
```

da-woods commented on docs/src/userguide/migrating_to_cy30.rst:
```diff
@@ -226,3 +226,16 @@ To make it easier to handle cases where your interpretation of type
+========================================
+
+Cython 3 differentiates between pre/post-increment and pre/pos-decrement
+operators (Cython 0.29 implemented both as predecrement operator).
```
```suggestion
operators (Cython 0.29 implemented both as pre(in/de)crement operator).
```

Not sure if this phrasing is clearer but it kind of implied that Cython 0.29 subtracted one whatever you did.

da-woods commented on tests/errors/cpp_preincrement.pyx:
```diff
@@ -0,0 +1,26 @@
+cimport cython
+
+cdef extern from *:
+    cdef cppclass Foo:
```
Could cpp_preincrement and cpp_postincrement be combined into one test? (I think it should be possible if the errors appear at the same stage). Largely because (I think) one bigger test file is quicker to run than two smaller test files.

If not possible it's fine though.

maxbachmann replied:
I merged them

da-woods reviewed (commented):
> It would be good to fix https://github.com/cython/cython/issues/4539 first.

I think we're prepared to ignore this for now right?

This PR seems a much simpler fix, so is probably worth going in even without the full non member bug fix.

maxbachmann commented:
> I think we're prepared to ignore this for now right?

Yes I would ignore this issue for now

maxbachmann commented:
@da-woods I applied the requested changes

da-woods commented:
Thanks @maxbachmann
